### ISSUE TITLE
Allow for multiple `tempo`

### ISSUE THREAD

**Issue by tmhglnd** [tmhglnd/mercury#94]
A feature that allows for setting tempo in multiple places would be interesting to give more freedom in sequencing.

For example if one would want to make a piece inspired by Piano Phase from Steve Reich one would have to make two instruments playing the same phrase but one is slightly faster then the other. This could look something like:

```js
list phrase random(16)

set tempo 120 
new synth sine note(phrase 1) time(1/16) // uses tempo 120 bpm

set tempo 121
new synth sine note(phrase 1) time(1/16) // uses tempo 121 bpm
```

This however requires a rework of mercury where instruments use separate clock sources instead of a global transport. Another option could be to allow for some tempo/time multiplication factor within the `time()` method, for example as a third argument.

```js
list phrase random(16)

set tempo 120
new synth sine note(phrase 1) time(1/16) // uses tempo 120 bpm

new synth sine note(phrase 1) time(1/16 0 1.00833333) //uses tempo 120 but multiplied by 1.008 to reach 121, the 0 is the offset argument that we don't use
```

**Comment by totalgee** [comment]
How does the clock/tempo work in Mercury when you re-evaluate code, does it restart at zero each time, or just keep running from where it was?

Also, if you're scaling/offsetting a single global clock, you need to decide what should happen on re-evaluation. Do you mind if it jumps in time, or do you want it to continue from the beat where it was, but now with the new tempo (or scale factor)?

**Comment by tmhglnd** [comment]
Thanks for joining the discussion! These are some good points to think about.

Currently the global transport of Mercury doesn't reset when you re-evaluate, it just continues where it was. In the browser only when you hit `alt+.` it resets to 0, and in Max only when the audio/dsp is turned-off and back on.

How I think of it now is that when rescaling the `time()` it would "jump" to the moment where it would be if that scaling was already effective since the beginning of the transport. This way I can make sure that when removing the scaling factor (or setting it back to `1`) the `time()` is back in sync with all the other instruments.

An additional feature could be that clocks can be desynchronised, allowing for rescaling from the moment of evaluation, but with the risk that things stay out of sync when the scaling goes back to `1`.

**Comment by tmhglnd** [comment]
An other experiment I tried was adding a `tempo()` method to all the instruments. This method is the global tempo (`set tempo`) by default. But if you add a different tempo, the ratio between that tempo and the global tempo is calculated, and the `time()` is adjust by that ratio.

Conceptually this could/should work, but in the browser version I encountered the limitation that not all floatingpoint values seem to give a good result. This is (I think) due to the fact that there is still one Global Transport in ToneJS, and it has a 192 PPQ (Pulses Per Quarternote) quantisation. So probably the only way to have multiple tempi in the browser version is to allow for multiple transports, or maybe a higher resolution PPQ (for example MaxMSP uses 480 ticks per quarternote)

**Comment by totalgee** [comment]
Yeah, interesting, it becomes a kind of sampling or aliasing issue because of that temporal quantization of that main clock (PPQ at its tempo). Yes, if only a discrete number of sub-steps are available and there is a single transport, then I don't think it could really be correct in general (for certain multiples it could be fine).

**Comment by tmhglnd** [comment]
Yeah indeed. I guess I should look into some other ways of scheduling the sequencer so it can be more accurate than the PPQ of ToneJS. Ideally it should be samplerate accurate for example. This is also how the sequencer for the mercury version in MaxMSP works, allowing for much more accurate control of the scheduled events. Something to look into. I also came across this article from Felix Roos discussing scheduling in the browser when he was developing Strudel. https://loophole-letters.vercel.app/web-audio-scheduling
